### PR TITLE
Initial Azure Power Control implementation - Revised

### PR DESCRIPTION
**Pull request by jteehan**
This is the short version now that most of the provisioning stuff is already in master.

{{pytest: cfme/tests/cloud/test_instance_power_control.py -k "test_stop" --long-running --use-provider azure }}

### CODE REVIEW

**Comment by jteehan** [comment]
Local Results:  Appliance's streams: [5.6, downstream]

cfme/tests/cloud/test_instance_power_control.py .

===================================================== 10 tests deselected by "-ktest_stop -m 'not perf'" =====================================================
========================================================= 1 passed, 10 deselected in 752.40 seconds ==========================================================

**Comment by dajoRH** [comment]
Lint report for commit 2a72a25ebc057798e6132a7ee39bafa1e8003470:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 60f019f04374271c94c16374aaa74a1a6c634df0:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit e59233f60e6a300d1af76eb0c8c82bd33a73f40c:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by psav** [comment]
Hey, all looks good except from PRT, I'm thinking perhaps a yaml update is needed?
```
ERROR:cfme:'region_ui'
Traceback (most recent call last):
  File "/cfme_tests_te/fixtures/provider.py", line 42, in _setup_provider
    providers.setup_provider(provider_key)
  File "/cfme_tests_te/utils/providers.py", line 257, in setup_provider
    provider.create(validate_credentials=True)
  File "/cfme_tests_te/cfme/common/provider.py", line 190, in create
    fill(self.properties_form, self._form_mapping(True, **self.__dict__))
  File "/cfme_tests_te/cfme/cloud/provider/azure.py", line 22, in _form_mapping
    'region_select': self.provisioning['region_ui'],
KeyError: 'region_ui'
```

**Comment by jteehan** [comment]
Good catch psav.  I think we copied from a slightly too early version and reverted back to a previous yaml value.  The yaml is fine and I pushed the correct value a minute ago.  As I may have mentioned, I need to regions in azure for the same region.  region_gui, which would look like "US East", and region_api which would look like "eastus" for the exact same region.  Goofy, but this fixes it.  Anyway, change was made.

**Comment by dajoRH** [comment]
Lint report for commit 6116f9022f69be065c530b4ed7b4c3621363e0d4:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by jteehan** [comment]
I was scratching my head trying to figure out why everything local passed but you got that error.  It's because my appliance already had the provider so it didn't need to add it.  Hence, it never got to the region section of the code.  I'll have to watch that going forward.

**Comment by dajoRH** [comment]
Lint report for commit 2ef31140e65b848ccf6c70257320417b04eab804:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by psav** [comment]
Problem has been found, 5.6.1 has a different region name

**Comment by psav** [comment]
How did we resolve this in the end?

**Comment by psav** [comment]
I see the change in the yamls for the region string, this will now fail on 55, if this is intended then we can add a skip for 55

**Comment by dajoRH** [comment]
Lint report for commit 1a0758a78e54320d0530ed70dbf7500ff9953333:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 6701f7560bf2b473ead4ac4c2e826667b72745eb:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 486f7afc77371a499cf7c13dc20f568a1950a91e:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 87d6cef57f6a187a2e5b9bdfa8b9bbe40d6eb18c:

Everything seems all right :smile: .
*CFME QE Bot*

File: cfme/cloud/instance.py [hunk 1]
```diff
@@ -367,7 +368,12 @@ class AzureInstance(Instance):
     START = "Start"
     POWER_ON = START  # For compatibility with the infra objects.
     STOP = "Stop"
-    TERMINATE = "Terminate"
+    SUSPEND = "Suspend"
```
Comment (jteehan) [inline cfme/cloud/instance.py]: So, this is either terminate or delete depending on the version.

File: cfme/cloud/provider/azure.py [hunk 2]
```diff
@@ -8,20 +10,33 @@ class AzureProvider(Provider):
...
         self.subscription_id = subscription_id
 
     def _form_mapping(self, create=None, **kwargs):
-        # Will still need to figure out where to put the tenant id.
-        return {'name_text': kwargs.get('name'),
-                'type_select': create and 'Azure',
-                'region_select': kwargs.get('region'),
-                'azure_tenant_id': kwargs.get('tenant_id'),
-                'azure_subscription_id': kwargs.get('subscription_id')}
+        data_dict = {
```
Comment (jteehan) [inline cfme/cloud/provider/azure.py]: This is the first time I've done one of these.  Need some eyes on it.  It works, but you may have a different format

File: cfme/common/provider.py [hunk 3]
```diff
@@ -66,36 +66,43 @@ class Credential(cfme.Credential, Updateable):
         """
         @property
         def form(self):
-            fields = [
-                ('token_secret_55', Input('bearer_token')),
-                ('google_service_account', Input('service_account')),
-            ]
-            tab_fields = {
-                "Default": [
+            if self.domain == 'azure' and version.current_version() >= '5.6.1':
```
Comment (jteehan) [inline cfme/common/provider.py]: This is what I came up with to fix the unique azure implementation.  I couldn't find provider.type or its equivelant and didn't want to add way to much code to find it.

File: cfme/common/provider.py [hunk 4]
```diff
@@ -675,7 +678,7 @@ def _fill_credential(form, cred, validate=None):
     elif cred.type == 'service_account':
         fill(cred.form, {'google_service_account': cred.service_account, 'validate_btn': validate})
     else:
-        if cred.domain:
+        if cred.domain and cred.domain != 'azure':
```
Comment (jteehan) [inline cfme/common/provider.py]: Workaround for the above workaround.

**Comment by dajoRH** [comment]
Lint report for commit 35a49b2e1cf5a138f8119ade1d0d97f5f7f23822:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 09e67778079fe5b38f2b41d8a65d583c91e5675c:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 9e261b462d8ba8844129f21eee28bcf4d31f243c:

`cfme/common/vm.py`:
- :red_circle: Line 620:70: **F821** *undefined name 'provider'*

`cfme/tests/cloud/test_instance_power_control.py`:
- :red_circle: Line 135:5: **F821** *undefined name 'request'*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 5a8197e2e6054e5e032f45d70d6d9e4b0174cd2d:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by jteehan** [comment]
We need to get this back into the review column and move it along.  I've implemented all the changes from Milan as requested.

**Pull request merged**